source/m3_env.c

```diff
@@ -806,6 +806,58 @@ _       ((M3Result) Call (i_function->compiled, (m3stack_t) stack, runtime->memo
     _catch: return result;
 }
 
+M3Result  m3_CallProper  (IM3Function i_function, uint32_t i_argc, const uint64_t* i_argv, unsigned *o_hasRet, uint64_t* o_ret)
```

Comment:
I think o_ret should be `void *`, otherwise the caller can't directly get an f32 value returned.

i.e.
```c++
float ret;
unsigned has_ret;
m3_CallProper(fn, argc, argv, &has_ret, (uint64_t*) &ret);
```
isn't valid. Instead they need to do
```c++
uint64 ret_qword;
float res;
unsigned has_ret;
m3_CallProper(fn, argc, argv, &has_ret, &ret_qword);
memcpy(&ret, &ret_qword, sizeof(ret));
```
which leaks the information about internal stack representation to the API.

Proposed:
```c++
m3_CallProper  (IM3Function i_function, uint32_t i_argc, const uint64_t* i_argv, void* o_ret, size_t ret_size);

...

float res;
m3_CallProper(fn, argc, argv, &res, sizeof(res));
```

Internally, m3_CallProper checks if ret_size is correct based on the function signature, and does a `memcpy` from the stack into `o_ret`.

Reply:
By the way, I also suggest changing `argv` to an array of `void*`. Reason is the same: how do you pass a bunch of f32 arguments into a function with the current API? User needs to do something like:
```c++
float a = ...;
float b = ...;

uint64_t args[2] = {};
memcpy(args[0], &a, sizeof(a));
memcpy(args[1], &b, sizeof(b));
m3_CallProper(fn, 2, args, ...);
```
Which, again, forces the caller to know about the internal stack representation.

Instead consider
```c++
m3_CallProper  (IM3Function i_function, uint32_t i_argc, const void* i_argv[], ...);

float a = ...;
float b = ...;

void* args[] = {&a, &b};
m3_CallProper(fn, sizeof(args)/sizeof(args[0]), args, ...);
```